lib/plugins/nwc/src/persist.rs

```diff
@@ -26,30 +43,180 @@ impl Persister {
+                .get_item(KEY_NWC_LOCK)?
+                .is_some_and(|lock| lock == "1");
+            if !is_locked {
+                break;
+            }
+            tokio::time::sleep(Duration::from_millis(500)).await;
+        }
+        if is_locked {
+            return Err(NwcError::generic("Could not acquire database lock"));
+        }
+        self.storage.set_item(KEY_NWC_LOCK, 1.to_string())?;
+        Ok(Lock { p: self })
+    }
+
+    pub(crate) fn unlock(&self) -> NwcResult<()> {
+        Ok(self.storage.set_item(KEY_NWC_LOCK, 0.to_string())?)
+    }
+
+    pub(crate) async fn add_nwc_connection(
+        &self,
+        name: String,
+        connection: NwcConnection,
+    ) -> NwcResult<()> {
+        let _ = self.lock().await?;
+        let mut connections = self.list_nwc_connections_inner()?;
+        if connections.contains_key(&name) {
+            return Err(NwcError::generic(format!(
+                "Could not insert connection: `{name}` already exists"
+            )));
+        }
+        connections.insert(name, connection);
         self.storage
-            .set_item(KEY_NWC_URIS, serde_json::to_string(&nwc_uris)?)?;
+            .set_item(KEY_NWC_CONNECTIONS, serde_json::to_string(&connections)?)?;
         Ok(())
     }
 
-    pub(crate) fn list_nwc_uris(&self) -> NwcResult<HashMap<String, String>> {
-        let raw_uris = self
+    // Helper method to update a connection's used budget directly
+    pub(crate) async fn update_periodic_budget(
+        &self,
+        name: &str,
+        periodic_budget: PeriodicBudget,
+    ) -> NwcResult<()> {
+        self.edit_nwc_connection(name, None, None, Some(periodic_budget))
+            .await?;
+        Ok(())
+    }
+
+    pub(crate) async fn edit_nwc_connection(
+        &self,
+        name: &str,
+        expiry_time_sec: Option<u32>,
+        receive_only: Option<bool>,
+        periodic_budget: Option<PeriodicBudget>,
+    ) -> NwcResult<NwcConnection> {
+        let _ = self.lock().await?;
+        let mut connections = self.list_nwc_connections_inner()?;
+        let Some(connection) = connections.get_mut(name) else {
+            return Err(NwcError::generic("Connection not found."));
+        };
+        if let Some(new_expiry_time) = expiry_time_sec {
+            connection.expiry_time_sec = Some(new_expiry_time);
+        }
+        if let Some(new_receive_only) = receive_only {
+            connection.receive_only = new_receive_only;
+        }
+        if let Some(new_periodic_budget) = periodic_budget {
+            connection.periodic_budget = Some(new_periodic_budget);
+        }
+        let connection = connection.clone();
+        self.storage
+            .set_item(KEY_NWC_CONNECTIONS, serde_json::to_string(&connections)?)?;
+        Ok(connection)
+    }
+
+    pub(crate) async fn get_min_interval(&self) -> Option<u64> {
+        let get_min_connection_interval = async || -> NwcResult<Option<u64>> {
+            let _ = self.lock().await?;
```

Comment:
I think the race concern is then read/writing, so I don't think need to hold a lock when reading